bgpd/bgp_mplsvpn.c

```diff
@@ -1336,6 +1336,225 @@ leak_update(struct bgp *to_bgp, struct bgp_dest *bn,
 	return new;
 }
 
+void bgp_mplsvpn_path_nh_label_unlink(struct bgp_path_info *path)
+{
+	struct bgp_label_per_nexthop_cache *blnc;
+
+	if (!path)
+		return;
+
+	blnc = path->label_nexthop_cache;
+
+	if (!blnc)
+		return;
+
+	LIST_REMOVE(path, label_nh_thread);
+	path->label_nexthop_cache->path_count--;
+	path->label_nexthop_cache = NULL;
+
+	if (LIST_EMPTY(&(blnc->paths)))
+		bgp_label_per_nexthop_free(blnc);
+}
+
+static int bgp_mplsvpn_get_label_per_nexthop_cb(mpls_label_t label,
+						void *context, bool allocated)
+{
+	struct bgp_label_per_nexthop_cache *blnc = context;
+	mpls_label_t old_label;
+	int debug = BGP_DEBUG(vpn, VPN_LEAK_LABEL);
+	struct bgp_path_info *path;
+	struct bgp_dest *dest;
+	struct bgp_table *table;
+
+	old_label = blnc->label;
+
+	if (debug)
+		zlog_debug("%s: label=%u, allocated=%d, nexthop=%pFX", __func__,
+			   label, allocated, &blnc->nexthop);
+	if (allocated)
+		/* update the entry with the new label */
+		blnc->label = label;
+	else
+		/*
+		 * previously-allocated label is now invalid
+		 * eg: zebra deallocated the labels and notifies it
+		 */
+		blnc->label = MPLS_INVALID_LABEL;
+
+	if (old_label == blnc->label)
+		return 0; /* no change */
+
+	/* update paths */
+	if (blnc->label != MPLS_INVALID_LABEL)
+		bgp_zebra_send_nexthop_label(ZEBRA_MPLS_LABELS_ADD, blnc->label,
+					     ZEBRA_LSP_BGP, &blnc->nexthop);
+
+	LIST_FOREACH (path, &(blnc->paths), label_nh_thread) {
+		dest = path->net;
+		if (!dest)
+			continue;
+		table = bgp_dest_table(dest);
+		if (!table)
+			continue;
+		vpn_leak_from_vrf_update(blnc->to_bgp, table->bgp, path);
+	}
+
+	return 0;
+}
+
+/* get label for a given bgp update
+ * return the label value:
+ * - per-vrf value:
+ *      if nexthop of path is not valid or there are no nexthops
+ *      if nexthop of path is not connected
+ *      if path becomes ECMP
+ *      if blackhole is found
+ * - valid per-nexthop value
+ *      if nexthop of path is valid and connected and a label has been obtained
+ * - invalid value
+ *      if nexthop of path is valid and connected and label has not yet been
+ * obtained
+ */
+static mpls_label_t bgp_mplsvpn_get_label_for_path_vrf(struct bgp_path_info *pi,
+						       struct bgp *to_bgp,
+						       struct bgp *from_bgp,
+						       afi_t afi)
+{
+	struct bgp_nexthop_cache *bnc = pi->nexthop;
+	struct bgp_label_per_nexthop_cache *blnc;
+	struct bgp_label_per_nexthop_cache_head *tree;
+	struct prefix *nh_to_use_ptr;
+	struct prefix nh_from_bnc = {0};
+
+	if (!bnc || !bnc->nexthop_num)
+		return from_bgp->vpn_policy[afi].tovpn_label;
+
+	/* check that the number of available nexthop is 1
+	 * ECMP is not supported
+	 * blackhole routes are not supported
+	 */
+	if (bnc->nexthop_num > 1 ||
+	    bnc->nexthop->type == NEXTHOP_TYPE_BLACKHOLE) {
+		/* if we are here, the bnc is valid, but does
```

Comment:
please no useless expression like `if we are here, `
Put the comment about ECMP and blackhole here